src/Compilers/CSharp/Portable/Compilation/SyntaxAndDeclarationManager.LazyState.cs

```diff
@@ -22,6 +21,15 @@ internal sealed class State
             internal readonly ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> LoadDirectiveMap;
             internal readonly ImmutableDictionary<string, SyntaxTree> LoadedSyntaxTreeMap;
             internal readonly ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> RootNamespaces;
+
+            /// <summary>
+            /// The last fully computed member names for all the types (in lexicographic order) for this file. Member
+            /// names often don't change for most edits, so being able to reuse the same set from the last time things
+            /// were computed saves on a lot of memory churn producing the new set, then GC'ing the last set (esp. for
+            /// very large types). The value is stored as a <see cref="OneOrMany"/> as the most common case for most
+            /// files is a single type declaration.
+            /// </summary>
+            internal readonly ImmutableDictionary<SyntaxTree, OneOrMany<ImmutableSegmentedHashSet<string>>> LastComputedMemberNames;
```

Comment:
>internal readonly ImmutableDictionary<SyntaxTree, OneOrMany<ImmutableSegmentedHashSet<string>>> LastComputedMemberNames; [](http://example.com/codeflow?start=12&length=120)

Would it make sense to use a weak reference here, so that it is collected when not used for a "long" time?

Reply:
Yes. That seems very reasonable.

Reply:
Ah, unfortunately ImmutableSegmentedHashSet is a struct, so i can't weakref that.  Personally, i don't view this value as particularly problematic.  It's in effect just a set of pointers to useful data in some prior compilation that thsi compilation depended on.  Most of the time all the values will be up to date, as we only really get LazyState for rapid edits of a file, so we would only have one of hte values be slightly stale (and only up to the point that we pull on the compilation next).

Reply:
> nfortunately ImmutableSegmentedHashSet is a struct, so i can't weakref that.

This doesn't sound to me like a good enough reason to hold on to the memory indefinitely. I am sure there are ways to overcome the "struct" obstacle.

Reply:
Switched to WeakRef to a StrongBox wrapper.  Tested and validated we still see the performance benefit (so in the typing case, we're still seeing the old data to reuse.).